### ISSUE TITLE
SET_SUBMIT_SUCCEEDED called from empty onSubmit and after stopSubmit with errors

### ISSUE THREAD

**Issue by jedrichards** [erikras/redux-form#2509]
I'm not returning a Promise from my `onSubmit`, instead dispatching an action to be later picked up in a saga.

The saga calls, `startSubmit('formName')`, and then at some point later calling `stopSubmit('formName', errors)` (or just `stopSubmit('formName')` if everything was ok).

In the case of a failure, the `errors` are being properly picked up and applied to the form, however, I'm seeing the `@@redux-form/SET_SUBMIT_SUCCEEDED` action coming through as well, which results in an inconsistent form state. The submit failed, but `@@redux-form/SET_SUBMIT_SUCCEEDED` is setting `submitSucceeded` to true even though I signalled a failed submission.

I'm trying to signal a failure, but the action sequence I'm getting is,

![screen shot 2017-01-26 at 12 04 25](https://cloud.githubusercontent.com/assets/1078571/22330731/9edf7328-e3bf-11e6-8b64-c0f5864cd2d5.png)

What's more, I get the `@@redux-form/SET_SUBMIT_SUCCEEDED` action creeping in even if I do nothing at all in `onSubmit`, i.e. with an empty `onSubmit` function that does nothing and doesn't even return a Promise I see the following action,

![screen shot 2017-01-26 at 13 02 52](https://cloud.githubusercontent.com/assets/1078571/22332240/cca74c42-e3c7-11e6-9a3a-efa26a65a2f5.png)

In terms of correctness, correct me if I'm wrong, but I feel like this action shouldn't be being called after an empty `onSubmit` that doesn't return a Promise. In that case shouldn't nothing have happened? A Promise wasn't returned, and `startSubmit()` wasn't called so why is Redux Form flipping `submitSucceeded` to `true`?

**Comment by Alxandr** [comment]
@jedrichards not all submission handlers are asynchronous. Not returning anything means things went OK (as opposed to an exception). What should probably be possible was to signal that an async action has started someway.

**Comment by jedrichards** [comment]
Agree. Redux Form is in the business of accurately modelling a form's state over time, but returning a promise being the only way to signal that a submit handler is async (or handled elsewhere e.g. in a saga) feels like quite a restrictive api. I wonder if a `onSubmitAsync` or something could be useful?

**Comment by Alxandr** [comment]
I've been experimenting with writing my own `reduxForm` function (while keeping the rest of `redux-form`) called `reduxSagaForm`. This seems to be the best solution to get perfect saga support.

**Comment by jedrichards** [comment]
I feel like there's _almost_ 100% perfect saga support as is (and indeed redux-thunk support, since it suffers from the same issue).

I setup my form like so,

```js
const withForm = reduxForm({
  onSubmit(values, dispatch) {
    dispatch(submit(values))
  },
  ...
})
```

and then whether the `submit` action creator is handled exclusively by a thunk, or triggers a saga, the approach is similar. The logic (in this case a saga) looks like,

```js
export function* submitWorker({ values }) {
  yield put(startSubmit('myForm'))
  yield put(someApiCall(values))

  const { success } = yield race({
    success: take('API_CALL_SUCCESS'),
    failure: take('API_CALL_FAILURE'),
  })

  if (failure) {
    yield put(stopSubmit('myForm', { _error: 'Failed' }))
  } else {
    // Handle success
  }
}
```

So remoting controlling the form like this with action creators basically _almost_ already works 100%, and I don't see how you'd achieve better separation of concerns. The problem being just the (fairly minor) inconsistency in the form's state that I mentioned at the top of this thread with `submitSucceeded` flipped to `true` - since as you rightly pointed out the submit handling logic is erroneously assumed synchronous.

This feels like something that could be fixed?

**Comment by jedrichards** [comment]
@Alxandr Your `reduxSagaForm` function sounds interesting, is this something you've published?

**Comment by Alxandr** [comment]
@jedrichards no cause it's just something I threw together for my own project. It doesn't support async validation for instance (though that should be fairly easy to add). I've uploaded a gist for you too look at: https://gist.github.com/Alxandr/b75f3c57246c8164bc412a74dc1942ea

It's not tested much or anything, but I couldn't use any promise "hacks" because I send my actions to a reducer in another process (that runs my saga), so both action and state has to be JSON serializeable. Therefore creating an action with the resolve function from a promise was a no go. And as such I wrote this.

**Comment by griffinmichl** [comment]
+1 to this comment. I submit my forms via redux-observable, and I have this same issue. It works _almost_ perfectly. Only complaint is the SET_SUBMIT_SUCCEEDED action.

**Comment by nicomo1989** [comment]
+1

**Comment by jedrichards** [comment]
Trying to handle both sync and async submit workflows via a single `onSubmit` handler, and forking the two behaviours based on the return type from that function feels like a slightly fuzzy public api. Is there any benefit to be gained by having something more explicit? And afaiu returning a promise from this handler is quite hostile to approaches that use libs that fully abstract such workflows like redux-observable and redux-saga.

Perhaps an `onSubmit` for sync workflows that is sensitive to errors being thrown, and an `onSubmitAsync` (or whatever) for async workflows where all of the submit handling is expected to be handled remotely via action creators?

I've got time and a desire to work on this and prepare a PR if the maintainers think there's some benefit to be gained here ...

**Comment by nzjoel1234** [comment]
This would be a great feature.
It would also be useful if the `onSubmitSuccess` callback was called whenever the submitSucceed action has been dispatched.

**Comment by jpdesigndev** [comment]
I agree with @jedrichards. The simplicity of it is nice (when dealing with promises or sync behavior), but if you're dealing with sagas, or, in my case, redux-observable, I can't seem to find right right balance. Perhaps, simply reimplementing `handleSubmit` these other use cases is the way to go..?

**Comment by nicolas-maurice** [comment]
I agree with @jpdesigndev, from my point of view this issue is due to the `handleSubmit` function that does not allow developpers to decide whether or not to dispatch redux-form actions (using `startSubmit`, 'stopSubmit', 'setSubmitSuceeded', etc.).

Maybe a solution would be to implement a `handleSubmitSilent` function that replicates `handleSubmit` behavior but does not dispatch redux-form actions giving developpers full power to handle these dispatchs in more abstracted workflows. (e.g. using `redux-saga`)

**Comment by erikras** [comment]
## Proposal

A new function, like `handleSubmit`, called `dispatchSubmit`, that you use like this:
```jsx
<form onSubmit={this.props.dispatchSubmit}>
  ...
</form>
```
It will check that sync validation and async validation all passes, and if and only if the form is valid, it will dispatch an action that looks something like:
```jsx
{
  type: '@@redux-form/DISPATCH_SUBMIT', // 'SUBMIT' is taken already... 😕
  payload: { 
    // all your form values
  }
}
```
The `redux-form` reducer will ignore this action entirely. It will be up to all your fancy async middlewares to dispatch `startSubmit()`, `stopSubmit()`, etc. to update the form's submitting state manually.

Would that satisfy everyone's needs?

**Comment by jedrichards** [comment]
That sounds great 🔥 

How would that interact with the `submitSucceeded` flag? As I mentioned at the start of this issue that seems to be getting flipped when `handleSubmit` exits. Presumably when using the proposed `dispatchSubmit` it should be being set conditionally when `stopSubmit` is called with/without errors?

**Comment by griffinmichl** [comment]
That sounds really great, @erikras. It's truly amazing how flexible this library can be while also doing so much work you.

**Comment by Alxandr** [comment]
@erikras one problem: how do we do async validation using redux-saga?

**Comment by griffinmichl** [comment]
I've managed to get async validation via redux-observable working, which should be the same idea. I've been meaning to make a PR to the FAQ section to show how I did it.

**Comment by erikras** [comment]
> How would that interact with the `submitSucceeded` flag? 

@jedrichards All those flags would be up to your middleware. For that particular flag, it would be set by dispatching the `setSubmitSuceeded` action.

> one problem: how do we do async validation using redux-saga?

@Alxandr Right, async validation is also promise based... I'm curious to hear what @griffinmichl has come up with.

**Comment by nicolas-maurice** [comment]
@erikras I think your suggestion to implement a `dispatchSubmit` function is great !! To go one step further, instead of always dispatching  the same `DISPATCH_SUBMIT` action, I would give possibility to dispatch any desired action.

To do so, a possible solution would be to enable developers to connect an action creator function as an optional property of the `reduxForm` decorator. The `dispatchSubmit` function you evoked would then use this action creator in order to create the action to dispatch.

It would something like
`reduxForm({
  form: "myForm",
  actionToDispatchOnSubmit: (values, props) => (actionCreator(values, props))
})(FormComponent)`

What do you think ?

Thanks for all the great work on this library !!

**Comment by jpdesigndev** [comment]
> It will be up to all your fancy async middlewares to dispatch startSubmit(), stopSubmit(), etc.

@erikras I like you're proposal, but my experience with this lib is quite limited (a few days). The **etc.** part is a bit concerning to me as I'm not 100% certain I could do it correctly without a good deal of knowledge about this lib. Perhaps a complete example using redux-saga / redux-observable would ease my concerns. I'm looking forward to hearing more from other folks here!

@griffinmichl I'm also curious to see what you've come up with. Let us know if you are able to get a PR up.

**Comment by burabure** [comment]
@erikras we really need something like your proposal! currently all of our form errors/validations are broken because we're using redux saga + redux form ;___;

**Comment by mthrm** [comment]
Being able to tell redux-form to simply dispatch an action to initiate form submission (and async validation) as @nicolas-maurice suggests would be great. That would allow us to remove the ugly hacks we currently have in place to make redux-form play nice with redux-saga.

**Comment by burabure** [comment]
an API like this https://github.com/colinbate/redux-form-submit-saga but integrated into redux form would be awesome

```js
const submit = onSubmitActions({
  submit: UPDATE_CUSTOMER_ADDRESS_START,
  success: UPDATE_CUSTOMER_ADDRESS_SUCCESS,
  failure: UPDATE_CUSTOMER_ADDRESS_FAILURE,
})

return (
  <form onSubmit={handleSubmit(submit)}>
  ...
```

**Comment by sebinsua** [comment]
> I've managed to get async validation via redux-observable working, which should be the same idea. I've been meaning to make a PR to the FAQ section to show how I did it.

@griffinmichl Can you share this example? I'm considering the same approach in my new job.

**Comment by istvano** [comment]
Hi Guys,

Any progress on this? I just started using redux-form and I have run into the same issue. I have found a sort of way to handle it, but I am not sure it is the best way. it looks like a little hack to me. If someone is interested it is this library: 
https://github.com/salsita/redux-form-actions

this takes advantage of the fact onSubmit can handle a promise as well.

I think it would be a much cleaner approach if one could define which action the form 
- triggers when a submit happens
- the form listens on failure and success events to update it's status

Thanks

**Comment by DJTB** [comment]
I'm also facing this issue. Using actions to handle async submission with `redux-logic`, similar to the above-mentioned `redux-saga` use cases.

@erikras your `dispatchSubmit` proposal looks like it would provide the functionality to handle the form easily.

**Comment by marcovdijk** [comment]
This proposal would be perfect for my current project. Any news on this?

**Comment by rootkc** [comment]
How is this going?

**Comment by gustavohenke** [comment]
Not going at all, folks. If you have the time to tackle this, please go ahead and send us a PR.

**Comment by superbiche** [comment]
In case it's useful, here is my workaround.  
I just return a Promise, and handle resolution/rejection in the main `submitFormSaga`:  

## Action creator
```javascript
export const ADDRESS_SUBMIT_REQUEST = 'ADDRESS_SUBMIT_REQUEST'
export const addressSubmitRequest = (form, values, promise) => ({
  type: ADDRESS_SUBMIT_REQUEST,
  form,
  values,
  promise
})
```

## Form 
```javascript
const onSubmit = (values, dispatch, props) => new Promise((resolve, reject) => {
  dispatch(addressSubmitRequest(
    // submitFormSaga needs to know which form to submit...
    props.form,
    // ... and which values to submit the form with
    {
      ...values,
      // These are additional settings allowing me to abstract my AddressForm, nothing interesting in this scope, but still useful to think of
      id: props.addressId,
      type: props.addressType,
      saveUser: props.onSubmitSaveUser,
    },
    // Pass the "promise" (actually an object containing the resolve and reject methods)
    { resolve, reject },
  ))
})

const AddressForm = props => (
  <form 
      className={classNames('AddressForm', props.className)} 
      onSubmit={props.handleSubmit}
  >
     // ...
  </form>

export default compose(
  connect((state, props) => ({
    ...props,
  })),
  reduxForm({
    validate,
    onSubmit,
  }),
)(AddressForm)
```

## sagas/form.js
```javascript
// This saga was greatly inspired by snippets found on issues regarding RF with sagas

export function* submitFormSaga(form, submitSaga, submitSagaArgs, promise) {
  const {resolve, reject} = promise
  // Start submitting, won't trigger setSubmitSucceeded()
  yield put(startSubmit(form))
  try {
    // Call the submitSaga, which takes care of the actual API flow etc.
    yield call(submitSaga, submitSagaArgs)
    // Allow resetting the form with a simple param
    if (submitSagaArgs && submitSagaArgs.resetForm) {
      yield put(reset(form))
    }
    // I was using this line before, but actions happened in a way I couldn't manage
    // Now resolving/rejecting the Promise will take care of dispatching stopSubmit()
    // yield put(stopSubmit(form))
    // Resolve the promise, dispatching setSubmitSucceeded()
    resolve()
  } catch (err) {
    // Handle errors depending on type
    let errors = null
    if (err instanceof SubmissionError) {
      error = err.errors
      // Same as above, resolving/rejecting the Promise will take care of dispatching stopSubmit()
      // yield put(stopSubmit(form, err.errors))
    } else {
      error = { _error: err }
      // yield put(stopSubmit(form, { _error: err }))
      logException(err)
    }
    // Finally reject the promise.
    reject(error)
  }
}
```

FWIW, here's the watcher saga (which calls submitAddressForm which itself calls `submitFormSaga`) and the data submit saga (called inside `submitFormSaga`):  
```javascript
// Watcher
export function* watchAddressSubmit() {
  yield takeLatest(userActions.ADDRESS_SUBMIT_REQUEST, submitAddressForm)
}
// Submit handler
export function* submitAddressForm(action) {
  yield call(submitFormSaga, action.form, saveAddress,
    // saveAddress arguments
    {
      form: action.form,
      ...action.values,
    },
    // Promise passed from onSubmit dispatched action.
    action.promise,
  )
}

// Data handling saga
export function* saveAddress(data) {
  try {
    const addressable_id = yield getAddressableId(data.type)

    const address = {
      ...data,
      addressable_type: data.type,
      addressable_id,
    }
    const response = yield call(Api.user.saveAddress, address)
    // Here we can call more sub-sagas depending on the parameters we've passed
    if (data.saveUser) {
      let userData = {
        ...data,
        id: yield select(getUserId),
      }
      yield call(saveUser, userData)
    }
    yield put(userActions.addressSubmitSuccess(response))
  } catch (e) {
    yield put(userActions.addressSubmitFailure(e))
    logException(e)
  } finally {
    // Mark the generator as "done"
    return true
  }
}
```


In result, here's the actions dispatched, in the expected order:  

![screen shot 2018-03-06 at 17 19 46](https://user-images.githubusercontent.com/2478146/37044055-bce5dae2-2162-11e8-8e7b-f1f34327e818.png)

This might not be the most straightforward way of doing it, but at least :  
- I can fully manage RF Forms from my Sagas  
- I don't have to adapt my codebase to what `redux-form-actions` or other packages force you to do  
- My RF form handling feature is only located in a single saga for the RF related part. Even if I had to adapt my sagas to handle form submissions + their Promises, once you get it, it's really easy to reproduce  
- As my logic is almost totally abstracted in my sagas, I can re-use my AddressForm by passing it different props depending on the use case, without needing to implement new sagas if the level of complexity stays reasonable

**Comment by erikras** [comment]
For all of you using `redux-saga`, perhaps [check out this tool I just built](https://medium.com/@erikras/getting-a-promise-from-redux-actions-d92787894368) to convert actions into a `Promise` for libraries like `redux-form`.

**Comment by ecnaidar** [comment]
@erikras This looks interesting, though still feels a lot like hack and adds complexity. If there is still interest, I might get a PR going for the `dispatchSubmit`.

**Comment by csantiago132** [comment]
@ecnaidar please!! ive been trying to connect redux-form to saga and theres nothing out there that seems to work

**Comment by csantiago132** [comment]
@superbiche do you have a full working example of your solution? Ive been struggling with this for three weeks and im almost there!

**Comment by ecnaidar** [comment]
Here is the PR for this behavior https://github.com/erikras/redux-form/pull/4015
All existing tests are passing and changes should be backwards compatible. Would be great to hear opinions on the matter and if there are any pitfalls I am unaware of. If everything is OK in approach and implementation, create tests for added functionality and ready to go.

**Comment by IRus** [comment]
So I'm another one who tried to do form submission in "redux way" (my middleware is redux-observable). I found proposed solutions hard to understand and maintain, but [original one](https://github.com/erikras/redux-form/issues/2509#issuecomment-294281646) is really what we need. 

Since I have just a few relatively simple forms, I ended up with rewriting code with formik and own redux integration. Integration is not really hard, and because formik uses TypeScript - it integrates better with project written in TypeScript.

I wish someday redux-form can be used with pure redux approach without pain, and replaced flow with TypeScript.

**Comment by Faolain** [comment]
ecnaidar has this solved here https://github.com/erikras/redux-form/pull/4015 . Seems like we just need this PR merged.

**Comment by pybuche** [comment]
Yay ! An estimated release date ? :)

**Comment by dimaqq** [comment]
Meanwhile I'm completely happy with:

```js
const save => () => form_props.handleSubmit(async () => {
  try {
    await api.foo();  // performs series of store.dispatch() with business data
  } catch (e) {
    log.error("...", e)
    throw new SubmissionError({foo: e.bar});
  }
})();
```

I think this can be adapted to saga's, etc., in the worst case by iterating saga manually.

**Comment by enoughtobesimple** [comment]
I am encountering a similar issue in React Native. Here, on a Login form when I press the submit button, it fires the SET_SUBMIT_SUCCEEDED action prior the LOGIN_SUCCESS. It doesn't navigate to the next screen until I press the Submit Button again (and with DESTROY this time). I am not sure if this is the expected behavior. I am using onSubmitSuccess function  for navigation to the next screen, and it happens on the second Button press though the form submission has passed in the first Button press. Could you point me in the right direction.

![ReduxFormSubmitButton](https://user-images.githubusercontent.com/53738000/66365669-6c781800-e9ab-11e9-968d-d1ae7a18fd80.PNG)

**Comment by dimaqq** [comment]
Since #4015 was merged, maybe this issue can be closed?
